### PR TITLE
Send basic contact to Emarsys via the API

### PR DESCRIPTION
**Pull request by timcwebster**
Sending basic contact data to Emarsys bia their API.

### CODE REVIEW

File: features/bootstrap/ApiContext.php [hunk 1]
```diff
@@ -71,4 +73,25 @@ public function iShouldReceiveASuccessfulEmarsysResponse()
...
         Assert::true($this->response->isSuccessful());
     }
+
+    /**
+     * @When I make a new customer API call with the following details
+     */
+    public function iMakeANewCustomerApiCallWithTheFollowingDetails(TableNode $table)
+    {
+        foreach ($table->getHash() as $data) {
+            $this->response = $this->application->addContact($data);
```
Comment (elvetemedve) [inline features/bootstrap/ApiContext.php:83]: Looks like you overwrite the response object here, so previous instances will be lost. Is that going to be a problem?

File: features/integration/api/create_contact.feature [hunk 2]
```diff
@@ -0,0 +1,10 @@
+@emarsys
+Feature: Ensure contacts are sent to Emarsys via the REST API
+  As a store owner
+  I want new customer dat to be sent to Emarsys via the API
```
Comment (elvetemedve) [inline features/integration/api/create_contact.feature]: Typo `dat`/`data`.

File: src/Inviqa/Emarsys/Api/Request/NewContactRequest.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        $this->client = $client;
+    }
+
+    /**
+     * @throws \LogicException
+     */
+    public function addContact(array $contactContent): ContactResponse
+    {
+        $body = [
+            'key_id' => '3', // This defaults to 3 (email) if not set, but setting anyway
```
Comment (elvetemedve) [inline src/Inviqa/Emarsys/Api/Request/NewContactRequest.php]: What does the 3 mean here? Maybe you should create a constant for it and give it a meaningful name.

File: src/Inviqa/Emarsys/Api/Request/NewContactRequest.php [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+use Inviqa\Emarsys\Api\Client;
+use Inviqa\Emarsys\Api\Response\ContactResponse;
+
+class NewContactRequest
+{
+    private $client;
+
+    private static $map = [
+        1, // First name
```
Comment (elvetemedve) [inline src/Inviqa/Emarsys/Api/Request/NewContactRequest.php]: I would create a private constant from it.

File: src/Inviqa/Emarsys/Api/Response/ContactResponse.php [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        $instance = new self();
+
+        $instance->isSuccessful = $response->isSuccessful();
+        $instance->validate($response);
+
+        return $instance;
+    }
+
+    public function isSuccessful()
```
Comment (elvetemedve) [inline src/Inviqa/Emarsys/Api/Response/ContactResponse.php]: Please add return type hint.

File: src/Inviqa/Emarsys/Api/Response/ContactResponse.php [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+namespace Inviqa\Emarsys\Api\Response;
+
+class ContactResponse
+{
+    private $isSuccessful;
+
+    /**
+     * @throws \LogicException
+     */
+    public static function fromClientResponse(ClientResponse $response): self
```
Comment (elvetemedve) [inline src/Inviqa/Emarsys/Api/Response/ContactResponse.php:12]: Why did you create a named constructor if there is only one?

File: src/Inviqa/Emarsys/Api/Response/ContactResponse.php [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    {
+        $instance = new self();
+
+        $instance->isSuccessful = $response->isSuccessful();
+        $instance->validate($response);
```
Comment (elvetemedve) [inline src/Inviqa/Emarsys/Api/Response/ContactResponse.php:17]: If the response is invalid, the `validate()` method will throw an exception, so the new instance will never be returned.
As a result `$instance->isSuccessful()` never will return false.

**Review by elvetemedve — CHANGES_REQUESTED** [review]

File: features/bootstrap/ApiContext.php (hunk 1 above)
Comment (timcwebster) [reply features/bootstrap/ApiContext.php:83]: no, not for the purposes of this PR, this is just very basic contact addition so will change in future PR

File: src/Inviqa/Emarsys/Api/Response/ContactResponse.php (hunk 6 above)
Comment (timcwebster) [reply src/Inviqa/Emarsys/Api/Response/ContactResponse.php:12]: This follows the same pattern as the other responses

File: src/Inviqa/Emarsys/Api/Response/ContactResponse.php (hunk 7 above)
Comment (timcwebster) [reply src/Inviqa/Emarsys/Api/Response/ContactResponse.php:17]: The Exception is caught in the Client, and a new response created there.

**Comment by timcwebster** [comment]
Am going to merge now, will be reviewed again when it's in a more complete state anyway

File: src/Inviqa/Emarsys/Api/Response/ContactResponse.php (hunk 7 above)
Comment (elvetemedve) [reply src/Inviqa/Emarsys/Api/Response/ContactResponse.php:17]: But that does not explain or solve the problem I mentioned.

**Pull request merged**